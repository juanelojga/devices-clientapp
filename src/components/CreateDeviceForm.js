import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { isEmpty, startCase, snakeCase, isNaN } from 'lodash';

import ConfirmationButton from './ConfirmationButton';

import styles from './CreateDeviceForm.module.scss';

import { DEVICE_TYPES } from '../configs/constants';

const CreateDeviceForm = ({
  system_name,
  type,
  hdd_capacity,
  handleSubmit,
  handleClose,
  loading
}) => {
  const [validationErrors, setValidationErrors] = useState({});
  const [systemName, setSystemName] = useState(system_name);
  const [systemType, setSystemType] = useState(snakeCase(type));
  const [hddCapacity, setHddCapacity] = useState(hdd_capacity);

  const onSubmit = e => {
    e.preventDefault();

    const capacity = parseInt(hddCapacity);
    // validation
    const errors = {};
    if (isEmpty(systemName)) errors['systemName'] = true;
    if (isEmpty(systemType)) errors['systemType'] = true;
    if (isEmpty(hddCapacity) || isNaN(capacity)) errors['hddCapacity'] = true;

    setValidationErrors(errors);

    if (isEmpty(errors)) {
      handleSubmit({
        system_name: systemName,
        type: systemType.toUpperCase(),
        hdd_capacity: capacity.toString()
      });
    }
  };

  return (
    <div>
      <h3>Device Form</h3>
      <form onSubmit={onSubmit} className={styles.form}>
        <div className={styles.group}>
          <input
            type="text"
            value={systemName}
            onChange={e => {
              setValidationErrors({});
              setSystemName(e.target.value);
            }}
            placeholder="System Name"
          />
          <label>System Name</label>
          {validationErrors['systemName'] ? (
            <p className={styles.error}>Invalid System Name</p>
          ) : null}
        </div>
        <div className={styles.group}>
          <select
            value={systemType}
            onChange={e => {
              setValidationErrors({});
              setSystemType(e.target.value);
            }}
          >
            <option value="" />
            {DEVICE_TYPES.slice(1).map(deviceType => (
              <option value={deviceType} key={deviceType}>
                {startCase(deviceType)}
              </option>
            ))}
          </select>
          <label>Type</label>
          {validationErrors['systemType'] ? (
            <p className={styles.error}>Invalid Type</p>
          ) : null}
        </div>
        <div className={styles.group}>
          <input
            type="text"
            value={hddCapacity}
            onChange={e => {
              setValidationErrors({});
              setHddCapacity(e.target.value);
            }}
            placeholder="HDD Capacity"
          />
          <label>HDD Capacity</label>
          {validationErrors['hddCapacity'] ? (
            <p className={styles.error}>Invalid HDD Capacity</p>
          ) : null}
        </div>
        <div className={styles.buttons}>
          <button onClick={handleClose} className="btn btn-red">
            Close
          </button>
          &nbsp;&nbsp;
          <ConfirmationButton
            type="submit"
            disabled={loading}
            loading={loading}
            text="Confirm"
          />
        </div>
      </form>
    </div>
  );
};

CreateDeviceForm.propTypes = {
  system_name: PropTypes.string,
  type: PropTypes.string,
  hdd_capacity: PropTypes.string,
  handleSubmit: PropTypes.func,
  handleClose: PropTypes.func,
  loading: PropTypes.bool
};

export default CreateDeviceForm;
