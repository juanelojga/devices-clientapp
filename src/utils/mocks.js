export const DEVICES = [
  {
    id: '1',
    system_name: 'SUSAN-DESKTOP',
    type: 'WINDOWS_WORKSTATION',
    hdd_capacity: '128 GB'
  },
  {
    id: '2',
    system_name: 'MAC-LOCAL-FREDDY',
    type: 'MAC',
    hdd_capacity: '256 GB'
  },
  {
    id: '3',
    system_name: 'SMART-SERVER',
    type: 'WINDOWS_SERVER',
    hdd_capacity: '1024 GB'
  }
];
