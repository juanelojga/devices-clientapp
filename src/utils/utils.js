export const filterByType = (filter, values) =>
  values.filter(value => (filter === 'all' ? true : value.type === filter));

export const sortBy = (criteria, values) =>
  values.sort((a, b) => {
    const criteria1 = a[criteria].toUpperCase();
    const criteria2 = b[criteria].toUpperCase();
    return criteria1.localeCompare(criteria2, undefined, {
      numeric: true,
      sensitivity: 'base'
    });
  });
