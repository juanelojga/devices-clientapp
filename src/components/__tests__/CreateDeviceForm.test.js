import React from 'react';
import { render, fireEvent, cleanup, getByText } from '@testing-library/react';
import 'jest-dom/extend-expect';

import CreateDeviceForm from '../CreateDeviceForm';

import { DEVICES } from '../../utils/mocks';

afterEach(cleanup);

test('snapshot test empty form', () => {
  const handleClose = jest.fn();
  const handleSubmit = jest.fn();
  const loading = false;
  const systemName = '';
  const type = '';
  const hddCapacity = '';

  const { container } = render(
    <CreateDeviceForm
      handleClose={handleClose}
      handleSubmit={handleSubmit}
      loading={loading}
      system_name={systemName}
      type={type}
      hdd_capacity={hddCapacity}
    />
  );
  expect(container).toMatchSnapshot();
});

test('snapshot test ', () => {
  const handleClose = jest.fn();
  const handleSubmit = jest.fn();
  const loading = false;
  const systemName = DEVICES[0].system_name;
  const type = DEVICES[0].type;
  const hddCapacity = DEVICES[0].hdd_capacity;

  const { container } = render(
    <CreateDeviceForm
      handleClose={handleClose}
      handleSubmit={handleSubmit}
      loading={loading}
      system_name={systemName}
      type={type}
      hdd_capacity={hddCapacity}
    />
  );
  expect(container).toMatchSnapshot();
});

test('not submit values because of validation errors', () => {
  const handleClose = jest.fn();
  const handleSubmit = jest.fn();
  const loading = false;
  const systemName = '';
  const type = '';
  const hddCapacity = '';

  const { getByText } = render(
    <CreateDeviceForm
      handleClose={handleClose}
      handleSubmit={handleSubmit}
      loading={loading}
      system_name={systemName}
      type={type}
      hdd_capacity={hddCapacity}
    />
  );

  fireEvent.click(getByText('Confirm'));
  expect(getByText('Invalid System Name')).toBeTruthy();
  expect(getByText('Invalid Type')).toBeTruthy();
  expect(getByText('Invalid HDD Capacity')).toBeTruthy();
  expect(handleSubmit).not.toHaveBeenCalled();
});

test('submit values', () => {
  const handleClose = jest.fn();
  const handleSubmit = jest.fn();
  const loading = false;
  const systemName = DEVICES[0].system_name;
  const type = DEVICES[0].type;
  const hddCapacity = DEVICES[0].hdd_capacity;

  const { getByText } = render(
    <CreateDeviceForm
      handleClose={handleClose}
      handleSubmit={handleSubmit}
      loading={loading}
      system_name={systemName}
      type={type}
      hdd_capacity={hddCapacity}
    />
  );

  fireEvent.click(getByText('Confirm'));
  expect(handleSubmit).toHaveBeenCalledTimes(1);
});
