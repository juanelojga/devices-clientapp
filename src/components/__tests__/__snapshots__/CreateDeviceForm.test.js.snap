// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test  1`] = `
<div>
  <div>
    <h3>
      Device Form
    </h3>
    <form
      class="form"
    >
      <div
        class="group"
      >
        <input
          placeholder="System Name"
          type="text"
          value="SUSAN-DESKTOP"
        />
        <label>
          System Name
        </label>
      </div>
      <div
        class="group"
      >
        <select>
          <option
            value=""
          />
          <option
            value="windows_workstation"
          >
            Windows Workstation
          </option>
          <option
            value="mac"
          >
            Mac
          </option>
          <option
            value="windows_server"
          >
            Windows Server
          </option>
        </select>
        <label>
          Type
        </label>
      </div>
      <div
        class="group"
      >
        <input
          placeholder="HDD Capacity"
          type="text"
          value="128 GB"
        />
        <label>
          HDD Capacity
        </label>
      </div>
      <div
        class="buttons"
      >
        <button
          class="btn btn-red"
        >
          Close
        </button>
          
        <button
          class="btn btn-green"
          type="submit"
        >
          Confirm
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`snapshot test empty form 1`] = `
<div>
  <div>
    <h3>
      Device Form
    </h3>
    <form
      class="form"
    >
      <div
        class="group"
      >
        <input
          placeholder="System Name"
          type="text"
          value=""
        />
        <label>
          System Name
        </label>
      </div>
      <div
        class="group"
      >
        <select>
          <option
            value=""
          />
          <option
            value="windows_workstation"
          >
            Windows Workstation
          </option>
          <option
            value="mac"
          >
            Mac
          </option>
          <option
            value="windows_server"
          >
            Windows Server
          </option>
        </select>
        <label>
          Type
        </label>
      </div>
      <div
        class="group"
      >
        <input
          placeholder="HDD Capacity"
          type="text"
          value=""
        />
        <label>
          HDD Capacity
        </label>
      </div>
      <div
        class="buttons"
      >
        <button
          class="btn btn-red"
        >
          Close
        </button>
          
        <button
          class="btn btn-green"
          type="submit"
        >
          Confirm
        </button>
      </div>
    </form>
  </div>
</div>
`;
